be/src/storage/compaction_manager.cpp

```diff
@@ -45,58 +49,161 @@ void CompactionManager::update_candidates(std::vector<CompactionCandidate> candi
+            return false;
+        }
+    } else if (candidate.type == CompactionType::BASE_COMPACTION) {
+        if (now_ms - last_failure_ts <= config::min_compaction_failure_interval_sec * 1000) {
+            VLOG(1) << "Too often to schedule failure compaction, skip it."
+                    << "compaction_type=" << starrocks::to_string(candidate.type)
+                    << ", min_compaction_failure_interval_sec=" << config::min_compaction_failure_interval_sec
+                    << ", last_failure_timestamp=" << last_failure_ts / 1000 << ", tablet_id=" << tablet->tablet_id();
+            return false;
+        }
+    }
+
+    return true;
 }
 
-CompactionCandidate CompactionManager::pick_candidate() {
+bool CompactionManager::pick_candidate(CompactionCandidate* candidate) {
     std::lock_guard lg(_candidates_mutex);
     if (_compaction_candidates.empty()) {
-        static CompactionCandidate invalid_candidate;
-        return invalid_candidate;
+        return false;
     }
 
     auto iter = _compaction_candidates.begin();
-    CompactionCandidate ret = *iter;
-    _compaction_candidates.erase(iter);
-    return ret;
+    while (iter != _compaction_candidates.end()) {
+        if (_check_precondition(*iter)) {
+            *candidate = *iter;
+            _compaction_candidates.erase(iter);
+            return true;
+        }
+        iter++;
+    }
+
+    return false;
+}
+
+void CompactionManager::_dispatch_worker() {
+    while (!_stop.load(std::memory_order_consume)) {
+        {
+            std::lock_guard lock(_dispatch_mutex);
+            if (!_dispatch_map.empty()) {
+                for (auto& [id, tablet_pair] : _dispatch_map) {
+                    auto& tablet = tablet_pair.first;
+                    Status st = _update_candidate_pool->submit_func([tablet, this] { update_tablet(tablet); });
+                    if (!st.ok()) {
+                        LOG(WARNING) << "update candidate tablet " << id << "failed. status:" << st.to_string();
+                    }
+                }
+                _dispatch_map.clear();
+            }
+        }
+        int32_t left_seconds = 10;
+        do {
+            sleep(1);
+            --left_seconds;
+        } while (!_stop.load(std::memory_order_consume) && left_seconds > 0 &&
+                 _max_dispatch_count < config::min_cumulative_compaction_num_singleton_deltas &&
+                 _dispatch_map.size() < 10240);
+    }
 }
 
-void CompactionManager::update_tablet_async(const TabletSharedPtr& tablet, bool need_update_context,
-                                            bool is_compaction) {
-    Status st = _update_candidate_pool->submit_func([tablet, need_update_context, is_compaction, this] {
-        update_tablet(tablet, need_update_context, is_compaction);
-    });
-    if (!st.ok()) {
-        LOG(WARNING) << "update candidate failed. status:" << st.to_string();
+void CompactionManager::update_tablet_async(TabletSharedPtr tablet) {
```

Comment:
Why not call it when add_rowset in tablet. It's the only entrance.

Reply:
not the only entrance, suck as clone not use that